### ISSUE TITLE
False positive from used-before-assignment when variable is reimported

### ISSUE THREAD

**Issue by MarcoGorelli** [PyCQA/pylint#7882]
### Bug description

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from foo import Bar

def myfunc():
    result: Bar

    from foo import Bar # if I remove this, then no error is reported

```

Note how if I remove the `from foo import Bar` line, then the error goes away

### Command used

```shell
pylint --disable=all --enable=used-before-assignment t.py
```

### Pylint output

```shell
************* Module t
t.py:7:12: E0601: Using variable 'Bar' before assignment (used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 1.67/10 (previous run: 1.67/10, +0.00)
```

### Expected behavior

No error reported

### Pylint version

```shell
pylint 2.15.7
astroid 2.12.13
Python 3.8.15 (default, Oct 12 2022, 19:15:16) 
[GCC 11.2.0]
```

**Comment by orSolocate** [comment]
Maybe it is an `astroid` issue and not Pylint?
Would be interesting to debug the `infer` used on `Bar` during `used-before-assignment` check..

**Comment by mbyrnepr2** [comment]
Thanks for the report. 

I think Pylint is struggling with this:
```python
if TYPE_CHECKING:
    from foo import Bar
```

Replacing with the following makes the false positive go away:
```python
from foo import Bar
```

**Comment by jacobtylerwalls** [comment]
Very similar to #7609, we need to account for the fact that annotations are not evaluated at runtime in the body of a function.

This works, but could use some clarifying comments:
```diff
diff --git a/pylint/checkers/variables.py b/pylint/checkers/variables.py
index 0d72ef969..e506009fd 100644
--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1799,9 +1799,14 @@ class VariablesChecker(BaseChecker):
                 # E0601 may *not* occurs in lambda scope.
 
                 # Handle postponed evaluation of annotations
-                if not (
-                    self._postponed_evaluation_enabled
-                    and isinstance(stmt, (nodes.AnnAssign, nodes.FunctionDef))
+                if not self._postponed_evaluation_enabled and (
+                    isinstance(stmt, nodes.FunctionDef)
+                    or (
+                        isinstance(stmt, nodes.AnnAssign)
+                        and not utils.get_node_first_ancestor_of_type(
+                            stmt, nodes.FunctionDef
+                        )
+                    )
                 ):
                     self.add_message(
                         "used-before-assignment",
```
